File: app/src/main/java/com/example/gemini_ai_chat/MainActivity.java

package com.example.gemini_ai_chat

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.gemini_ai_chat.databinding.ActivityMainBinding
import com.google.gson.Gson
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.BufferedReader
import java.io.InputStreamReader

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var chatAdapter: ChatAdapter
    private lateinit var db: AppDatabase
    private lateinit var geminiApi: GeminiApi
    private lateinit var keyRotator: KeyRotator

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val messages = mutableListOf<Message>()
        chatAdapter = ChatAdapter(messages)
        binding.chatRecyclerView.layoutManager = LinearLayoutManager(this)
        binding.chatRecyclerView.adapter = chatAdapter

        db = AppDatabase.getDatabase(this)
        geminiApi = RetrofitClient.getClient()
        keyRotator = KeyRotator(this)

        loadMessages(messages)

        binding.sendButton.setOnClickListener {
            val inputText = binding.messageInputText.text.toString()
            if (inputText.isNotEmpty()) {
                sendMessage(inputText, messages)
            }
        }
    }

    private fun loadMessages(messages: MutableList<Message>) {
        lifecycleScope.launch(Dispatchers.IO) {
            val loadedMessages = db.messageDao().getAllMessages()
            withContext(Dispatchers.Main) {
                messages.addAll(loadedMessages)
                chatAdapter.notifyDataSetChanged()
                binding.chatRecyclerView.scrollToPosition(messages.size - 1)
            }
        }
    }

    private fun sendMessage(text: String, messages: MutableList<Message>) {
        val userMessage = Message(text = text, isUser = true, timestamp = System.currentTimeMillis())
        addMessage(userMessage, messages)
        binding.messageInputText.setText("")

        lifecycleScope.launch(Dispatchers.IO) {
            val messageId = db.messageDao().insert(userMessage)
            userMessage.id = messageId.toInt()

            val aiMessage = Message(text = "", isUser = false, timestamp = System.currentTimeMillis())
            val aiMessageId = db.messageDao().insert(aiMessage)
            aiMessage.id = aiMessageId.toInt()

            withContext(Dispatchers.Main) {
                addMessage(aiMessage, messages)
            }

            val apiKey = keyRotator.getNextKey() ?: ""
            val request = GeminiModels.Request(text)
            geminiApi.streamGenerateContent(apiKey, request).enqueue(object : Callback<ResponseBody> {
                override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                    if (response.isSuccessful) {
                        lifecycleScope.launch(Dispatchers.IO) {
                            val reader = BufferedReader(InputStreamReader(response.body()?.byteStream()))
                            val responseBuilder = StringBuilder()
                            var line: String?
                            while (reader.readLine().also { line = it } != null) {
                                if (line!!.contains("\"text\":")) {
                                    val jsonResponse = line!!.substring(line!!.indexOf("{"), line!!.lastIndexOf("}") + 1)
                                    try {
                                        val geminiResponse = Gson().fromJson(jsonResponse, GeminiModels.Response::class.java)
                                        if (geminiResponse?.candidates?.isNotEmpty() == true) {
                                            val newText = geminiResponse.candidates[0].content.parts[0].text
                                            responseBuilder.append(newText)
                                            aiMessage.text = responseBuilder.toString()
                                            updateMessage(aiMessage, messages)
                                        }
                                    } catch (e: Exception) {
                                        // Ignore JSON parsing errors for incomplete lines
                                    }
                                }
                            }
                            db.messageDao().update(aiMessage)
                        }
                    } else {
                        aiMessage.text = "Error: ${response.code()}"
                        updateMessage(aiMessage, messages)
                        lifecycleScope.launch(Dispatchers.IO) {
                            db.messageDao().update(aiMessage)
                        }
                    }
                }

                override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
                    aiMessage.text = "Failure: ${t.message}"
                    updateMessage(aiMessage, messages)
                    lifecycleScope.launch(Dispatchers.IO) {
                        db.messageDao().update(aiMessage)
                    }
                }
            })
        }
    }

    private fun addMessage(message: Message, messages: MutableList<Message>) {
        runOnUiThread {
            messages.add(message)
            chatAdapter.notifyItemInserted(messages.size - 1)
            binding.chatRecyclerView.scrollToPosition(messages.size - 1)
        }
    }


    private fun updateMessage(message: Message, messages: MutableList<Message>) {
        runOnUiThread {
            val index = messages.indexOfFirst { it.id == message.id }
            if (index != -1) {
                messages[index] = message
                chatAdapter.notifyItemChanged(index)
            }
        }
    }
}

File: app/src/main/java/com/example/gemini_ai_chat/MainActivity.kt

package com.example.gemini_ai_chat

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Button
import android.widget.EditText
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.auth.FirebaseAuth

class MainActivity : AppCompatActivity() {

    private lateinit var messageViewModel: MessageViewModel
    private lateinit var messageAdapter: MessageAdapter
    private lateinit var auth: FirebaseAuth

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        auth = FirebaseAuth.getInstance()

        if (auth.currentUser == null) {
            val intent = Intent(this, LoginActivity::class.java)
            startActivity(intent)
            finish()
            return
        }

        setContentView(R.layout.activity_main)

        val recyclerView = findViewById<RecyclerView>(R.id.recycler_view)
        val inputEditText = findViewById<EditText>(R.id.input_text)
        val sendButton = findViewById<Button>(R.id.send_button)

        messageAdapter = MessageAdapter()
        recyclerView.adapter = messageAdapter
        recyclerView.layoutManager = LinearLayoutManager(this)

        messageViewModel = ViewModelProvider(this).get(MessageViewModel::class.java)
        messageViewModel.allMessages.observe(this, {
            messageAdapter.submitList(it)
        })

        sendButton.setOnClickListener {
            val inputText = inputEditText.text.toString()
            if (inputText.isNotEmpty()) {
                val message = Message(text = inputText, isUser = true, timestamp = System.currentTimeMillis())
                messageViewModel.insert(message)
                inputEditText.text.clear()
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_settings -> {
                val intent = Intent(this, SettingsActivity::class.java)
                startActivity(intent)
                true
            }
            R.id.action_logout -> {
                FirebaseAuth.getInstance().signOut()
                val intent = Intent(this, LoginActivity::class.java)
                startActivity(intent)
                finish()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}
